### PR TITLE
feat: v2 migration helpers

### PR DESCRIPTION
**Pull request by dummdidumm**
- svelte-migrate stuff
- migration docs

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [ ] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [ ] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [ ] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

File: documentation/docs/60-appendix/11-v2-migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+export function load({ fetch }) {
+-    const response = fetch(...).then(r => r.json());
++    const response = await fetch(...).then(r => r.json());
+    return { response }
+}
+```
+
+## paths are now relative by default
+
+TODO explain
```
Comment (dummdidumm) [inline documentation/docs/60-appendix/11-v2-migration-guide.md:43]: @Rich-Harris I need your help to explain what impact this has

File: documentation/docs/60-appendix/11-v2-migration-guide.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+## Top-level promises are no longer awaited
+
+In SvelteKit version 1, promises returned from the top level of a `load` function are automatically awaited. Very few people actually took advantage of this, and with the introduction of [streaming load functions](https://svelte.dev/blog/streaming-snapshots-sveltekit) this behavior became a bit awkward as it forces you to nest your streamed data one level deep. Therefore, SvelteKit no longer awaits top level promises returned from a `load` function and instead streams these, too. To get back the blocking behavior, wrap your top level promises with `Promise.all(...)`, or if you only have one promise, `await` it.
+
+```diff
+export function load({ fetch }) {
+    const response1 = fetch(...).then(r => r.json());
+    const response2 = fetch(...).then(r => r.json());
+-    return { response1, response2 }
++    return Promise.all({ response1, response2 })
```
Comment (Rich-Harris) [inline documentation/docs/60-appendix/11-v2-migration-guide.md:30]: `Promise.all` doesn't work this way

```suggestion
-    const a = fetch(...).then(r => r.json());
-    const b = fetch(...).then(r => r.json());
+    const [a, b] = Promise.all([
+      fetch(...).then(r => r.json()),
+      fetch(...).then(r => r.json()),
+    ]);
    return { a, b };
```

Comment (Rich-Harris) [reply documentation/docs/60-appendix/11-v2-migration-guide.md:30]: a diff view of a diff code block is quite a confusing sight to behold

File: documentation/docs/60-appendix/11-v2-migration-guide.md (hunk 1 above)
Comment (benmccann) [reply documentation/docs/60-appendix/11-v2-migration-guide.md:50]: I think it offers better support for things like the Internet Archive and IPFS

**Comment by benmccann** [comment]
I think we were a little inconsistent in the language of some of the changesets. Maybe we can align them as part of this. Especially `.changeset/gentle-guests-clean.md` where I said "drop support for Svelte 3" I think maybe we could word in a more upbeat way like "require Svelte 4 to ease the path for Svelte 5 migration" or something.

**Comment by benmccann** [comment]
I think it'd be nice for this to appear before "Migrating from Sapper" in the sidebar since it will be relevant to more users. And maybe we can call it "Migrating to SvelteKit v2" rather than "Migration to SvelteKit v2" for consistency, so that both entries begin with "Migrating"

File: packages/migrate/migrations/sveltekit-2/migrate.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+export function update_tsconfig() {
+	fs.writeFileSync(
+		'tsconfig.json',
+		update_tsconfig_content(fs.readFileSync('tsconfig.json', 'utf8'))
+	);
+}
+
+/** @param {string} content */
+export function update_tsconfig_content(content) {
```
Comment (benmccann) [inline packages/migrate/migrations/sveltekit-2/migrate.js:38]: can we also make sure typescript is on v5 and that `moduleResolution` is `bundler` for applications and `NodeNext` or `Node16` for libraries?

**Comment by Rich-Harris** [comment]
I wonder if we should merge this into `version-2` already — that way, other PRs can add migration notes and `svelte-migrate` stuff. e.g. I'd like to auto-migrate `resolvePath` but I want to do that in #11261, not here

**Comment by Rich-Harris** [comment]
> Maybe we can align them as part of this

I think that should happen on the `version-2` branch directly

> I think it'd be nice for this to appear before "Migrating from Sapper" in the sidebar since it will be relevant to more users. And maybe we can call it "Migrating to SvelteKit v2" rather than "Migration to SvelteKit v2" for consistency, so that both entries begin with "Migrating"

done

**Comment by dummdidumm** [comment]
Merging now and doing the other changes in their respective PRs sounds good.
Things I have in mind for migration that are part of already merged PRs, which could maybe added here before merging:
- add `path` to `cookies` methods where needed/possible
- more tsconfig tweaks like removing the moduleresolution and module if it's outdated settings (`node`)
- remove `dangerZone` from `svelte.config.js`

**Comment by Rich-Harris** [comment]
Am trying to insert `// @migration task: Specify path` comments and I'm just totally lost. Seems like a simple enough thing to do but I can't see any APIs for prepending text, only replacing (which fucks up indentation). Frankly I suspect we'd have a much better time if we ditched `ts-morph` in favour of using the same traversal techniques we use in `svelte` (i.e. using `acorn` with `acorn-typescript` to generate an `estree` AST which is much nicer to deal with than TypeScript insanity).

I've wasted enough time fighting with it and there are more important things to do, so I'm just doing to check in my work so far and merge this PR. Those other changes can be follow-ups.

File: documentation/docs/60-appendix/11-v2-migration-guide.md (hunk 2 above)
Comment (janvotava) [reply documentation/docs/60-appendix/11-v2-migration-guide.md]: Isn't that missing `await` before `Promise.all` and also `async` in the function header?

**Pull request merged**